### PR TITLE
mv: fix missing view deletions in some cases of range tombstones

### LINKED ISSUE #17117: Range tombstones can be missed in view if many deletions get sent to the same view partition
For efficiency, several view updates to the same view partition are batched together to one mutation. However, if this mutation becomes too big Scylla doesn't want to handle it all together, so it limits the size of each batch to 100 rows (this is the `max_rows_for_view_updates` constant in `view_update_builder::on_results()`). 

We noticed that **range tombstones** (i.e., deletion of clustering-key prefix) in the base update are counted incorrectly - they have multiple parts so seem to be counted multiple times and only 25 view-row deletions are generated instead of 100. Moreover, the batch can be ended _in the middle of_ a range tombstone, causing this range tombstone to not generate a view update.

The following test reproduces this issue. A table has 200 rows and a materialized view, and generates a batch of 200 base _range tombstones_ each covering one row. This should result in 200 row tombstones in the view. Since all view updates are to the same partition they are batched. But instead of generating batches of 100 deletions we see (in debug printouts) batches of 25 deletion. Worse - the 26th, 52nd, etc. deletion doesn't happen in the view. At the end of the test, the base is empty (we deleted all its rows), but the view still has its 26th, 52nd, etc. rows alive.

```
def test_many_range_tombstone_base_update(cql, test_keyspace):
    N = 200
    with new_test_table(cql, test_keyspace, 'p int, c1 int, c2 int, primary key (p, c1, c2)') as table:
        with new_materialized_view(cql, table, '*', 'p, c1, c2', 'p is not null and c1 is not null and c2 is not null') as mv:
            insert = cql.prepare(f'INSERT INTO {table} (p, c1, c2) VALUES (?,?,?)')
            p = 42
            for i in range(N):
                cql.execute(insert, [p, i, i])
            cmd = 'BEGIN BATCH '
            for i in range(N):
                cmd += f'DELETE FROM {table} WHERE p={p} AND c1={i} '
            cmd += 'APPLY BATCH;'
            cql.execute(cmd)
            # At this point, both base table and view tables should be
            # empty.
            assert [] == list(cql.execute(f'SELECT c1 FROM {table}'))
            assert [] == list(cql.execute(f'SELECT c1 FROM {mv}'))
```

The bug seems to stem from how `on_results()` counts the number of rows in the view update, using the function `op_count()`. This function appears to be buggy - instead of counting _just_ the number of rows, it additionally adds 1 every time it is called.  Beyond over-counting, the problem of this is that we increment the counter when we're in the middle of processing a range tombstone, and if we reached the counter 100 at that point, we stop this batch and lose the range tombstone we were processing.

The following on-line patch seems to fix this bug:
```patch
diff --git a/db/view/view.cc b/db/view/view.cc
index 5fe54eb03d..14a9a0f36e 100644
--- a/db/view/view.cc
+++ b/db/view/view.cc
@@ -491,7 +491,8 @@ mutation_partition& view_updates::partition_for(partition_key&& key) {
 }
 
 size_t view_updates::op_count() const {
-    return _op_count++;;
+    return _op_count;
 }
 
 row_marker view_updates::compute_row_marker(const clustering_or_static_row& base_row) const {
```

I don't know why this code was introduced like that in commit https://github.com/scylladb/scylla-enterprise/commit/bf0777e97a0abb3bebddf84c82de774ebfd9255d - was it always a bug? Is the above the correct fix?

I'm worried that this bug is just a tip of the iceberg... Could counting and stopping a batch at exactly count 100 cause additional problems of stopping in the middle of a batch (perhaps involving some deletion of an old row and creation of a new row), or this bug could only happen for range tombstones? The code in this area is super-convoluted and looks very delicate and I wonder how we can test it better, beyond the test above that focuses just on the range-tombstone case. Perhaps we can try BATCHes of other types of updates, or maybe even random combinations of different types of updates?

### PR DESCRIPTION
**Pull request by nyh**
For efficiency, if a base-table update generates many view updates that go the same partition, they are collected as one mutation. If this mutation grows too big it can lead to memory exhaustion, so since commit 7d214800d0ea3ec21d40ac7a10f2fbf2c11c6098 we split the output mutation to mutations no longer than 100 rows (max_rows_for_view_updates) each.

This patch fixes a bug where this split was done incorrectly when the update involved range tombstones, a bug which was discovered by a user in a real use case (#17117).

Range tombstones are read in two parts, a beginning and an end, and the code could split the processing between these two parts and the result that some of the range tombstones in update could be missed - and the view could miss some deletions that happened in the base table.

This patch fixes the code in two places to avoid breaking up the processing between range tombstones:

1. The counter "_op_count" that decides where to break the output mutation should only be incremented when adding rows to this output mutation. The existing code strangely incrmented it on every read (!?) which resulted in the counter being incremented on every *input* fragment, and in particular could reach the limit 100 between two range tombstone pieces.

2. Moreover, the length of output was checked in the wrong place... The existing code could get to 100 rows, not check at that point, read the next input - half a range tombstone - and only *then* check that we reached 100 rows and stop. The fix is to calculate the number of rows in the right place - exactly when it's needed, not before the step.

The first change needs more justification: The old code, that incremented _op_count on every input fragment and not just output fragments did not fit the stated goal of its introduction - to avoid large allocations. In one test it resulted in breaking up the output mutation to chunks of 25 rows instead of the intended 100 rows. But, maybe there was another goal, to stop the iteration after 100 *input* rows and avoid the possibility of stalls if there are no output rows? It turns out the answer is no - we don't need this _op_count increment to avoid stalls: The function build_some() uses `co_await on_results()` to run one step of processing one input fragment - and `co_await` always checks for preemption. I verfied that indeed no stalls happen by using the existing test test_long_skipped_view_update_delete_with_timestamp. It generates a very long base update where all the view updates go to the same partition, but all but the last few updates don't generate any view updates. I confirmed that the fixed code loops over all these input rows without increasing _op_count and without generating any view update yet, but it does NOT stall.

This patch also includes two tests reproducing this bug and confirming its fixed, and also two additional tests for breaking up long deletions that I wanted to make sure doesn't fail after this patch (it doesn't).

By the way, this fix would have also fixed issue #12297 - which we fixed a year ago in a different way. That issue happend when the code went through 100 input rows without generating *any* output rows, and incorrectly concluding that there's no view update to send. With this fix, the code no longer stops generating the view update just because it saw 100 input rows - it would have waited until it generated 100 output rows in the view update (or the input is really done).

Fixes #17117

### CODE REVIEW

**Comment by nyh** [comment]
> #### 🔴 CI State: FAILURE
> 

ְSeems like a new flaky test, unrelated to this PR :-( Opened https://github.com/scylladb/scylladb/issues/17167 and will re-run the CI.

**Comment by mykaul** [comment]
memory exhaustion protection is still there, as we continue to limit the /output/ (which is what actually consumes more and more memory?) I assume?

File: db/view/view.cc [hunk 1]
```diff
@@ -491,7 +491,7 @@ mutation_partition& view_updates::partition_for(partition_key&& key) {
 }
 
 size_t view_updates::op_count() const {
-    return _op_count++;;
+    return _op_count;
```
Comment (avikivity) [inline db/view/view.cc:494]: We can probably de-mutable view_updates::_op_count now.

Generally, adding mutable means "I'm hiding a bug here".

Comment (avikivity) [reply db/view/view.cc:494]: Ah you did that.

File: db/view/view.cc [hunk 2]
```diff
@@ -1308,11 +1308,12 @@ void view_update_builder::generate_update(static_row&& update, const tombstone&
...
     constexpr size_t max_rows_for_view_updates = 100;
-    size_t rows_for_view_updates = std::accumulate(_view_updates.begin(), _view_updates.end(), 0, [] (size_t acc, const view_updates& vu) {
-        return acc + vu.op_count();
-    });
-    const bool stop_updates = rows_for_view_updates >= max_rows_for_view_updates;
-
+    auto should_stop_updates = [this] () -> bool {
+        size_t rows_for_view_updates = std::accumulate(_view_updates.begin(), _view_updates.end(), 0, [] (size_t acc, const view_updates& vu) {
+            return acc + vu.op_count();
+        });
```
[the remark this answers is not in the document]
Comment (nyh) [reply db/view/view.cc:1314]: I just moved the existing code a few lines (the second bug I fixed was that we did this summation *before* adding a row, not after).
Personally, I was never a fan of this game of "Perl golf". I think even clearer and much easier to read is simply:
```C++
    size_t rows_for_view_updates;
    for (const auto& vu : _view_updates) {
       rows_for_view_updates += vu.op_count();
    }
```

**Comment by avikivity** [comment]
I don't fully understand the fix. I can understand a problem if on_results() is called in the middle of computing a range tombstone pair, causing only half the pair to be emitted. But where/how does that happen?

**Comment by avikivity** [comment]
Yes, the computation of _op_count was incorrect and unstable, but where does the decision wrt. closing a range tombstone change?

**Comment by nyh** [comment]
> memory exhaustion protection is still there, as we continue to limit the /output/ (which is what actually consumes more and more memory?) I assume?

Yes, exactly: There are still "op_count++" in all the functions that increase the size of the mutation we're collecting (adding a row, deleting a row), so when the output mutation reaches 100 rows, we stop the loop and send the mutation out. So we never keep a mutation longer than 100 rows in memory.

The **read** side memory limitation was also solved (long ago): Yes, we may need to iterate over a base partition of a million rows, but we never read that entire partition into memory - we use the mutation-reader-fragment machinary to read pieces of the input mutation one by one.

**Comment by avikivity** [comment]
btw we should change "100 ops" to "128k bytes" or something.

**Comment by nyh** [comment]
> I don't fully understand the fix. I can understand a problem if on_results() is called in the middle of computing a range tombstone pair, causing only half the pair to be emitted. But where/how does that happen?

I'm not sure I explained the problem right. Let me try again.

Basically the problem is that when `on_results()` returns `stop_iteration::yes`, the caller (`build_some()`) will return, and later `build_some()` will be called again to build the next 100 (or whatever) rows. But when `build_some()` is called, it doesn't **only** call on_results() - before calling it it calls `advance_all()`, which advances *both* the existing and update. Let's look at an example to see why it's wrong: Imagine you have:

Existing: A, B, C, D
Update: RA1, RA2, RB1, RB2, RC1, RC2, RD1, RD2

Where A is a row, and RA1 and RA2 are two pieces of a range tombstone that covers A.
The algorithm proceeds - when it works correctly, as follows (sort of):
1. `build_some()` advances the existing and update points to point to A and RA1, respectively.
2. `on_results()` is called, sees RA1 and A have the same position, and reads RA1, and advances update pointer to RA2
3. `on_results()` is called, sees RA2 and A have the same position, reads RA2, and advances update pointer to RB1
4. `on_results()` is called, sees RB1 is past A, so generates the saved tombstone for A. And advances the existing pointer to B. Only now that we generated an output, we increment op_count and may decide to return return from build_some().
6. and so on.

The bug happens when we return from build_some() not after step 4, but say after step 2. The problem isn't that we were middle of the range tombstone over A - the problem is that we never generated any output for A. As I mentioned the next call to build_some() will begin by advancing both existing and update pointers, so the existing pointer will move to B - the tombstone over A was never generated!

To fix this I needed to make sure we don't exit build_some() (i.e., don't return stop_iteration::yes) in steps 1..3 - only in step 4. The second bug I fixed (the position of the total op_count calculation) was even more ridiculous: At the end of step 4 we can exit, but we didn't - we did another step 5, and only then decided to exit based on the situation in step 4! In step 5, we moved to B, but then we exit the loop and lose B (because the next build_some() will move to C).

**Comment by nyh** [comment]
I admit this code is super convoluted, and I wouldn't be surprised if there are additional bugs left, even though we continue to expand our test suite with more and more elaborate situations. This is the second bug in exactly this logic that I have fixed after being found by a customer (the previous one https://github.com/scylladb/scylladb/issues/12297). But I am not sure what to do - I don't really want to rewrite all this code (which Durate wrote in d5a61a8c48261a3478527871927ae6ed01c22ed2, Piotr Sarna added the 100-row batching in 7d214800d0ea3ec21d40ac7a10f2fbf2c11c6098, and also at some point the range tombstones changed).

**Comment by nyh** [comment]
> > btw we should change "100 ops" to "128k bytes" or something.
> 
> This sounds like a follow-up idea?

Yes, this can probably be done fairly easy, but it's unrelated to the current bug. 
On the other hand, I'm not sure what exactly this will gain us. We need to read whole rows anyway, so if someone has a huge blob or something, does it really matter if we read one or 100 of them? (yes, 100 is more than 1, but we have a problem in either case, no?).

**Comment by nyh** [comment]
By the way, I didn't mention this explicitly, but one of the tests I wrote (included in the PR), one that does a batch of 234 range tombstones shows that both parts of my fix was necessary - before any patch it lost every 26th deletion (25 = 100 /4 - there were 4 on_results() steps in moving both updates and existing one position), after the op_count increment removal it lost every 101st deletion - every time we stopped the iteration one step too late. Only after both fixes, the test begins to work even for 234 deletions.

**Comment by nyh** [comment]
> 100 huge blobs is a lot worse than 1 huge blob

They are both O(1) ;-)

**Comment by avikivity** [comment]
> > 100 huge blobs is a lot worse than 1 huge blob
> 
> They are both O(1) ;-)

1 Megabyte in a shard is easy, 100 Megabytes is stressful. Both are O(1).

**Comment by nyh** [comment]
Opened https://github.com/scylladb/scylladb/issues/17188 on the 100-rows vs. size-in-bytes thing.

**Comment by mykaul** [comment]
@nyh - was it backported to 5.2 and 5.4? Can we remove the labels?

**Comment by nyh** [comment]
> @nyh - was it backported to 5.2 and 5.4? Can we remove the labels?

Is this how these tags are supposed to work? After you backport to 5.2 you should remove the backport/5.2 tag?
In the past, these tags remained behind, to remind us of the decision to backport. Only the "backport candidate" tag was removed (from the issue, not the PR).
Anyway, I'll remove them now.
